core/types/auxpow.go

```diff
@@ -0,0 +1,798 @@
+}
+
+func (ah *AuxPowHeader) SetNonce64(nonce uint64) {
+	if ah.inner == nil {
+		return
+	}
+	ah.inner.SetNonce64(nonce)
+}
+
+func (ah *AuxPowHeader) SetMixHash(mixHash common.Hash) {
+	if ah.inner == nil {
+		return
+	}
+	ah.inner.SetMixHash(mixHash)
+}
+
+func (ah *AuxPowHeader) SetHeight(height uint32) {
+	if ah.inner == nil {
+		return
+	}
+	ah.inner.SetHeight(height)
+}
+
+func (ah *AuxPowHeader) SetNonce(nonce uint32) {
+	if ah.inner == nil {
+		return
+	}
+	ah.inner.SetNonce(nonce)
+}
+
+func (ah *AuxPowHeader) setInner(inner AuxHeaderData) {
+	ah.inner = inner
+}
+
+func (ah *AuxPowHeader) Copy() *AuxPowHeader {
+	if ah.inner == nil {
+		return &AuxPowHeader{}
+	}
+	return &AuxPowHeader{inner: ah.inner.Copy()}
+}
+
+func NewAuxPowHeader(inner AuxHeaderData) *AuxPowHeader {
+	auxHeader := new(AuxPowHeader)
+	auxHeader.setInner(inner)
+	return auxHeader
+}
+
+type AuxPowTx struct {
+	inner AuxPowTxData
+}
+
+type AuxPowTxData interface {
+	Serialize(w io.Writer) error
+	SerializeNoWitness(w io.Writer) error
+	Deserialize(r io.Reader) error
+	DeserializeNoWitness(r io.Reader) error
+	Copy() AuxPowTxData
+}
+
+func NewAuxPowCoinbaseTx(powId PowID, height uint32, coinbaseOut []byte, auxMerkleRoot common.Hash, signatureTime uint32) []byte {
+	switch powId {
+	case Kawpow:
+		return NewRavencoinCoinbaseTx(height, coinbaseOut, auxMerkleRoot, signatureTime)
+	case SHA_BTC:
+		return NewBitcoinCoinbaseTx(height, coinbaseOut, auxMerkleRoot, signatureTime)
+	case SHA_BCH:
+		return NewBitcoinCashCoinbaseTx(height, coinbaseOut, auxMerkleRoot, signatureTime)
+	case Scrypt:
+		return NewLitecoinCoinbaseTx(height, coinbaseOut, auxMerkleRoot, signatureTime)
+	default:
+		return []byte{}
+	}
+}
+
+// NewAuxPowTx creates an AuxPowTx from an AuxPowTxData implementation
+func NewAuxPowTx(inner AuxPowTxData) *AuxPowTx {
+	return &AuxPowTx{inner: inner}
+}
+
+func (ac *AuxPowTx) Bytes() []byte {
+	if ac.inner == nil {
+		return nil
+	}
+	var buffer bytes.Buffer
+	ac.inner.SerializeNoWitness(&buffer)
+	return buffer.Bytes()
+}
+
+func (ac *AuxPowTx) Copy() *AuxPowTx {
+	if ac.inner == nil {
+		return &AuxPowTx{}
+	}
+	return &AuxPowTx{inner: ac.inner.Copy()}
+}
+
+func (ac *AuxPowTx) Serialize(w io.Writer, witness bool) error {
+	if ac.inner == nil {
+		return errors.New("inner transaction is nil")
+	}
+	if witness {
+		return ac.inner.Serialize(w)
+	}
+	return ac.inner.SerializeNoWitness(w)
+}
+
+func (ac *AuxPowTx) Deserialize(r io.Reader, witness bool) error {
+	if ac.inner == nil {
+		return errors.New("inner transaction is nil")
+	}
+	if witness {
+		return ac.inner.Deserialize(r)
+	}
+	return ac.inner.DeserializeNoWitness(r)
+}
+
+// AuxPow represents auxiliary proof-of-work data
+type AuxPow struct {
+	powID        PowID         // PoW algorithm identifier
+	auxPow2      []byte        // Auxiliary proof-of-work data
+	header       *AuxPowHeader // 120B donor header for KAWPOW
+	signature    []byte        // Signature proving the work
```

Comment:
This field is never checked directly right? It's always only checked after ConvertToTemplate?

Reply:
yeah true, only checked on the verify signature, but is part of the hash though